File: src/julia/runner/Benchmark.jl
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

module Benchmark
include("load.jl")
include("../shared/load.jl")
using ADPerfTest
using TestLoader
using GMMData
using BAData
using HandData
using LSTMData
using DataUtils

export run_benchmark

const measurable_time_not_achieved = -1

"""
    find_repeats_for_minimum_measurable_time!(context, minimum_measurable_time, func)

Finds the number of repeats that `func` needs to be asked to compute its objective
in order to to run for at least `minimum_measurable_time` seconds.
`func` must have a method `func(::typeof(context), repeats::Int)`. That is the method
that will be tested.

Returns a tuple `(repeats, min_sample, total_time)`, where
`repeats` is the found necessary number of repeats,
`min_sample` is the minimal encountered value of
`@elapsed func(context, repeats) / repeats`
`total_time` is the total time spent on running `func` in seconds.
"""
function find_repeats_for_minimum_measurable_time!(context::Any, minimum_measurable_time::Float64, func::Function)
    total_time = 0.0
    min_sample = floatmax(Float64)
    repeats = 1
    max_possible_power_of_two = typemax(Int) ÷ 2 + 1
    while true
        current_run_time = @elapsed func(context, repeats)
        if current_run_time > minimum_measurable_time
            current_sample = current_run_time / repeats
            min_sample = min(min_sample, current_sample)
            total_time += current_run_time
            break
        end
        # The next iteration will overflow a loop counter that's why we recognize that we cannot reach the minimum measurable time.
        if repeats == max_possible_power_of_two
            repeats = measurable_time_not_achieved
            break
        end
        repeats *= 2
    end
    repeats, min_sample, total_time
end

"Benchmarks func according to the methodology described in docs/Methodology.md"
function measure_shortest_time!(context::Any, minimum_measurable_time::Float64, nruns::Int, time_limit::Float64, func::Function)::Float64
    precompile(func, (typeof(context), Int))
    repeats, min_sample, total_time = find_repeats_for_minimum_measurable_time!(context, minimum_measurable_time, func)
    if repeats == measurable_time_not_achieved
        throw(ErrorException("It was not possible to reach the number of repeats sufficient to achieve the minimum measurable time."))
    end
    # Recount `time_limit` and `run` from 0, despite "find_repeats_for_minimum_measurable_time",
    # because there might be lazy intializations
    total_time = 0.0
    run = 0
    while run < nruns && total_time < time_limit
        current_run_time = @elapsed func(context, repeats)
        min_sample = min(min_sample, current_run_time / repeats)
        total_time += current_run_time
        run += 1
    end
    min_sample
end

"Performs the entire benchmark process according to the methodology described in docs/Methodology.md"
function run_benchmark(
    input::Input,
    input_name::AbstractString,
    module_name::AbstractString,
    output_prefix::AbstractString,
    module_display_name::AbstractString,
    minimum_measurable_time::Float64,
    nruns_f::Int,
    nruns_J::Int,
    time_limit::Float64
) where Input
    output = create_empty_output_for(input)
    test = get_test_for(input, output, module_name)
    # Callbacks defined in test exist only in the latest world
    Base.invokelatest() do
        test.prepare!(test.context, input)

        objective_time = measure_shortest_time!(test.context, minimum_measurable_time, nruns_f, time_limit, test.calculate_objective!)
        derivative_time = measure_shortest_time!(test.context, minimum_measurable_time, nruns_J, time_limit, test.calculate_jacobian!)

        test.output!(output, test.context)

        save_time_to_file(times_file_name(output_prefix, input_name, module_display_name), objective_time, derivative_time)
        save_output_to_file(output, output_prefix, input_name, module_display_name)
    end
end

end

File: src/python/runner/Benchmark.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

import time
import sys
import struct
from collections import namedtuple

from runner.ModuleLoader import module_load
from shared import GMMData
from shared import BAData
from shared import HandData
from shared import LSTMData
from runner.Filepaths import filepath_to_basename, modulepath_to_basename
from shared.output_utils import save_time_to_file

measurable_time_not_achieved = -1

def find_repeats_for_minimum_measurable_time(minimum_measurable_time, func):
    total_time = 0
    min_sample = sys.float_info.max

    repeats = 1
    # get platform C maxint & (maxint >> 1) + 1
    max_reasonable_number_of_repetitions = 2 ** (struct.Struct('i').size * 8 - 2)
    while True:
        t1 = time.time()
        func(repeats)
        t2 = time.time()
        # Time in seconds
        current_run_time = t2 - t1
        if current_run_time > minimum_measurable_time:
            current_sample = current_run_time / repeats
            min_sample = min(min_sample, current_sample)
            total_time += current_run_time
            break
        repeats *= 2
        # loop exit condition
        if repeats > max_reasonable_number_of_repetitions:
            # we recognize that we cannot reach the minimum measurable time.
            repeats = measurable_time_not_achieved
            break

    result = namedtuple('result', 'repeats, sample, total_time')
    return result(repeats = repeats, sample = min_sample, total_time = total_time)

# Measures time according to the documentation.
def measure_shortest_time(minimum_measurable_time, nruns, time_limit, func):
        find_repeats_result = find_repeats_for_minimum_measurable_time(minimum_measurable_time, func)

        if find_repeats_result.repeats == measurable_time_not_achieved:
            raise RuntimeError("It was not possible to reach the number of repeats sufficient to achieve the minimum measurable time.")

        repeats = find_repeats_result.repeats
        min_sample = find_repeats_result.sample
        total_time = find_repeats_result.total_time

        # "run" begins from 1 because a first run already done by "find_repeats_for_minimum_measurable_time" function
        run = 1
        while (run < nruns) and (total_time < time_limit):
            run += 1
            t1 = time.time()
            func(repeats)
            t2 = time.time()
            # Time in seconds
            current_run_time = t2 - t1
            min_sample = min(min_sample, current_run_time / repeats)
            total_time += current_run_time

        return min_sample

# Performs the entire benchmark process according to the documentation
def run_benchmark(module_path, input_filepath, _input, output_prefix, minimum_measurable_time, nruns_F, nruns_J, time_limit):

    test = module_load(module_path)

    test.prepare(_input)

    objective_time = measure_shortest_time(minimum_measurable_time, nruns_F, time_limit, test.calculate_objective)

    derivative_time = measure_shortest_time(minimum_measurable_time, nruns_J, time_limit, test.calculate_jacobian)

    output = test.output()

    input_basename = filepath_to_basename(input_filepath)
    module_basename = modulepath_to_basename(module_path)

    save_time_to_file(output_prefix + input_basename + "_times_" + module_basename + ".txt", objective_time, derivative_time)
    output.save_output_to_file(output_prefix, input_basename, module_basename)